File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/LmsFeatureModelsBehavior.kt
package com.lextures.android.core.lms

import kotlinx.serialization.Serializable

// region Behavior / PBIS (M10.3)

@Serializable
data class BehaviorCategory(
    val id: String,
    val orgId: String,
    val name: String,
    val type: String,
    val color: String? = null,
    val active: Boolean = true,
) {
    val isPositive: Boolean get() = type.equals("positive", ignoreCase = true)
    val isNegative: Boolean get() = type.equals("negative", ignoreCase = true)
}

@Serializable
data class BehaviorCategoriesResponse(
    val categories: List<BehaviorCategory> = emptyList(),
)

@Serializable
data class PbisAwardInput(
    val studentId: String,
    val categoryId: String,
    val points: Int = 1,
    val note: String? = null,
)

@Serializable
data class PbisAwardsBody(
    val awards: List<PbisAwardInput>,
)

@Serializable
data class PbisAward(
    val id: String,
    val studentId: String,
    val awardedBy: String? = null,
    val categoryId: String,
    val categoryName: String? = null,
    val orgId: String? = null,
    val points: Int,
    val note: String? = null,
    val awardedAt: String,
)

@Serializable
data class PbisAwardsResponse(
    val saved: Int? = null,
    val awards: List<PbisAward> = emptyList(),
    val message: String? = null,
)

@Serializable
data class BehaviorReferralBody(
    val studentId: String,
    val categoryId: String,
    val schoolId: String? = null,
    val incidentAt: String? = null,
    val location: String? = null,
    val description: String,
    val response: String? = null,
)

@Serializable
data class BehaviorReferral(
    val id: String,
    val studentId: String,
    val filedBy: String? = null,
    val orgId: String? = null,
    val schoolId: String? = null,
    val categoryId: String,
    val categoryName: String? = null,
    val incidentAt: String,
    val location: String? = null,
    val description: String? = null,
    val response: String? = null,
    val createdAt: String,
)

@Serializable
data class StudentBehaviorResponse(
    val studentId: String,
    val totalPoints: Int,
    val awards: List<PbisAward> = emptyList(),
    val referrals: List<BehaviorReferral> = emptyList(),
)

@Serializable
data class HallPass(
    val id: String,
    val sectionId: String,
    val studentId: String? = null,
    val destination: String,
    val status: String,
    val estimatedMins: Int? = null,
    val requestedAt: String,
    val approvedAt: String? = null,
    val returnedAt: String? = null,
    val approvedBy: String? = null,
    val overdue: Boolean? = null,
)

@Serializable
data class HallPassResponse(
    val pass: HallPass? = null,
)

@Serializable
data class ActiveHallPassesResponse(
    val passes: List<HallPass> = emptyList(),
)

@Serializable
data class RequestHallPassBody(
    val destination: String,
    val estimatedMins: Int,
)

@Serializable
data class UpdateHallPassBody(
    val status: String,
)

// endregion

File: clients/ios/Lextures/Core/LMS/LMSFeatureModelsBehavior.swift
import Foundation

// MARK: - Behavior / PBIS (M10.3)

struct BehaviorCategory: Codable, Identifiable, Hashable {
    var id: String
    var orgId: String
    var name: String
    var type: String
    var color: String?
    var active: Bool

    var isPositive: Bool { type.lowercased() == "positive" }
    var isNegative: Bool { type.lowercased() == "negative" }
}

struct BehaviorCategoriesResponse: Decodable {
    var categories: [BehaviorCategory]?
}

struct PBISAwardInput: Encodable {
    var studentId: String
    var categoryId: String
    var points: Int
    var note: String?
}

struct PBISAwardsBody: Encodable {
    var awards: [PBISAwardInput]
}

struct PBISAward: Codable, Identifiable, Hashable {
    var id: String
    var studentId: String
    var awardedBy: String?
    var categoryId: String
    var categoryName: String?
    var orgId: String?
    var points: Int
    var note: String?
    var awardedAt: String
}

struct PBISAwardsResponse: Decodable {
    var saved: Int?
    var awards: [PBISAward]?
    var message: String?
}

struct BehaviorReferralBody: Encodable {
    var studentId: String
    var categoryId: String
    var schoolId: String?
    var incidentAt: String?
    var location: String?
    var description: String
    var response: String?
}

struct BehaviorReferral: Codable, Identifiable, Hashable {
    var id: String
    var studentId: String
    var filedBy: String?
    var orgId: String?
    var schoolId: String?
    var categoryId: String
    var categoryName: String?
    var incidentAt: String
    var location: String?
    var description: String?
    var response: String?
    var createdAt: String
}

struct StudentBehaviorResponse: Decodable {
    var studentId: String
    var totalPoints: Int
    var awards: [PBISAward]?
    var referrals: [BehaviorReferral]?
}

// MARK: - Hall pass (M10.3)

struct HallPass: Codable, Identifiable, Hashable {
    var id: String
    var sectionId: String
    var studentId: String?
    var destination: String
    var status: String
    var estimatedMins: Int?
    var requestedAt: String
    var approvedAt: String?
    var returnedAt: String?
    var approvedBy: String?
    var overdue: Bool?
}

struct HallPassResponse: Decodable {
    var pass: HallPass?
}

struct ActiveHallPassesResponse: Decodable {
    var passes: [HallPass]?
}

struct RequestHallPassBody: Encodable {
    var destination: String
    var estimatedMins: Int
}

struct UpdateHallPassBody: Encodable {
    var status: String
}
